### ISSUE TITLE
Output file of graphics only

### ISSUE THREAD

**Issue by perpetersen** [ijprest/keyboard-layout-editor#130]
Would it be possible to have an output file of the graphics only in high res pdf? If so the graphics would be printable on a physical keyboard and a great help for those that want to design their own one-and-only keyboard, ex for an application specific shortcut keyboard.
![img_140326](https://cloud.githubusercontent.com/assets/14123458/9682673/37517668-5309-11e5-8886-1e9822d38423.JPG)

**Comment by iandoug** [comment]
Does it matter if the printout does not look like a keyboard? I take it you are going to print stickers or something (perhaps for dye-sub?)? Or are you looking for something you can feed to a laser etcher / whatever?

Suggested font size?

thanks.

**Comment by perpetersen** [comment]
Hi Ian
Thank you for comming back to me. Let me give you a bit of background. I am representing a keyboard manufacturer and we would like to offer a solution to make personalized keyboards, meaning you would be able to design your own keyboard on an editor like this, and we would do the physical keyboard and send it to you. We would print each keycap with the customer specified background color, texts and icons. To keep complexity to a minimum, we would need to offer this solution based on our standard series of keyboards, so as to the way the printout needs to look, we would have very strict critereas as the output design needs to be compatible with our keyboard design, and the output format needs to be compatible with our printing equipment... Perhaps it would be easier for me to explain all of this to you in detail over a Skype call? (I am located in Denmark, GMT +1)

**Comment by iandoug** [comment]
Yes, I have browsed your website before so I have a reasonable idea of what you do. Your keycaps are quite complex. What sort of technology are you using to print them (if that's not a trade secret)?. I assume you print they keys before they get assembled into the keyboard? I'm trying to figure out if your process involves printing the whole key (and/or keyboard) in one go, or if the keys are done separately, possibly in two stages ... top and front separately or together?  I'm guessing judging from your description above that you're going to load all the keys in some sort of holder/frame and then print them all at once? So you'll need a precisely-laid-out PDF?
BTW I'm just a contributor, not in charge ... that's the other Ian. :-)

**Comment by perpetersen** [comment]
Hi Ian, I am sorry but I can’t reveal our production methods here out in the open. What I can confirm is that we would need a precisely-laid-out PDF, or perhaps other file format depending on what is possible... As we have a commercial interest in this, we are of cause considering investing in the project. Do you know how I can get in contact with the other Ian?

**Comment by iandoug** [comment]
If a PDF, what would the page size need to be set at? I have been exploring how to do this, but have two 'issues', one of which is page size. For example my "programmer's keyboard" just fits on an A3, but I don't think 'standard' 104 layouts will.. just a bit too wide (although possibly just the key legends will). The other is the font ... we would need to embed the font in the document, and that may bring up licence issues. For example the default font that KLE uses for some key profiles is hosted on the live site but is not in the git repo because it is a proprietary font ... I'm not sure what the licence allows. Of course if you are running your own version locally using local fonts then there probably won't be a problem.

I'm sure Ian Prest will respond to you in due course, he was busy this weekend and it is a long weekend in the USA/Canada today.

**Comment by perpetersen** [comment]
You are right that the page size needs to be bigger than A3. Format would be something like 460 X 160 mm. Would it be a problem with such odd size output? Regarding the font, I believe we would just use one font and not make it optional to select between a variety of fonts. Is the font used on the Layout Editor today paid for or how is that handled? As to the keyboard template, is it designed in millimeters or in pixels? What resolution can it be defined in (decimals)?

**Comment by iandoug** [comment]
I was hoping IJP would be back by now. Anyway it looks like your size will fit on a B3 page. Which may raise other issues when it comes to printing. It looks like it may be possible to specify a page size by dimensions rather than ISO name when creating a PDF, which again may raise issues when it comes to printing. The KLE uses various fonts, some of which are distributed with various OS (Windows/Mac), as well as Engraver's Gothic, which is the problematic one. IJP uses this as a close match to Signature Plastics' "Gorton (modified)" font, which does not exist in an electronic form. (http://www.keycapsdirect.com/faq.php#15). If the user does not have the specified font installed, then the browser will use whatever it can, typically what the user specified as the default sans or sans-serif font. KLE uses two 'custom' self-made fonts, one for the Combining Diacriticals and the other for the WebFonts (special characters), as well as FontAwesome. I don't think there are any issues with any of those.
As for measurements, IJP  will be better to answer here ... I see he has set up both pixels and millimetres but I'm not sure where he uses what. You can take a look at render.js, he is working to 4 decimals.

**Comment by iandoug** [comment]
@perpetersen: is it possible to supply us with your colour codes so that we can add a swatch (or 2) for your colour schemes? Thanks.

**Comment by ijprest** [comment]
I'm back now (was a long weekend in N.A.). I haven't had much time to digest this, but a few thoughts:

* Currently, the easiest way to get just the labels is to select all the keys and check the 'decal' checkbox.
  * It would be straightforward to have a 'treat-all-as-decals' mode for download/export.
* The current output (even to SVG) is done in pixel-based units, but I've done some prototyping exporting SVG in millimeters (which is why the 'mm' options exist in the code).
* The biggest problem with vector-based (SVG/PDF) output is that mimicking the browser's text/font handling is **very** difficult.  
  * I haven't attempted it yet for SVG.  SVG output is currently just the keycap *borders*... exactly the part you don't want!  :)
* An easier solution might be to export as a high-resolution bitmap... 300 dpi or something, especially since @iandoug already figured out how to export bitmaps at different sizes.
  * Biggest problem might be performance / memory usage.

Fonts:
* I've licensed a couple for the website, as @iandoug mentioned.  
  * [Volkswagen Serial](http://www.fontspring.com/fonts/softmaker/volkswagen-serial) -- for the *Apple* look
  * [Engraver's Gothic FS](http://www.fontspring.com/fonts/fontsite/engravers-gothic-fs) -- for the *Signature Plastics* look.
* Neither font is perfect, or even the best available, but the licensing terms from fontspring.com are easier for a small-time open-source project to deal with than the more well-known font providers (which tend to require page tracking scripts and the like).
  * I'm not an expert in font licensing, but you might need an 'Application' license for something like this.

**Comment by iandoug** [comment]
I fiddled my code that does the thumbnails and went in the other direction, here's a 300DPI image. I was actually wrong about my design fitting on an A3, only the keyplate does. So my graphics programs increase the density to 334 DPI to force it to fit on an A3 (largest printer I have).
Can @perpetersen perhaps give an opinion? Can you print it at 300DPI? It's the full keyboard not just the legends, but I can probably get just the legends using the approach IJP gave above. 
Issue: legends on the front of the keys are not going to print properly using his approach, only the legends on the top of the keys.
It does however bypass any font issues... :)
However since we are upscaling an in-memory image, the fonts do get a bit blurry. Will have to rather find a way of getting the legends created in hi-res in the first place.
Sorry this is png, ended up rather large.
![keyboard-hires](https://cloud.githubusercontent.com/assets/12673886/9798605/4c5b5ec2-5802-11e5-9a47-af85d9adf6e6.png)

**Comment by iandoug** [comment]
I added a button to the Tools to do a CSS zoom to 300 DPI, which works, but when you try the save-as-jpg or save-as-png you get a mess. So I think the correct approach will be to redraw the layout using different CSS that will scale the fonts appropriately for a 300 DPI image.... after removing the decals and background, and making all the other keys decals.

**Comment by perpetersen** [comment]
Dear Ian & Ian
I am currently on business in Amsterdam for the IBC show, so I am sorry for my late reply, but I will have to make it short for now... I truly admire your expertise and your commitment to this project, but it starts to get a little too technical for my knowledge about this. What I do understand is, that this is not just a walk in the park thing to do. But let me get back to you later when I have more time… IJP, is there an email or something where I can get in contact with you outside this forum?

**Comment by iandoug** [comment]
I had an idea which may be a workaround.
0. set all the keys to decals, so that we don't see them.
1. as above, I zoomed to 300DPI using some javascript and CSS.
2. print the keyboard... but selected "print to PDF" which is built into Linux ... Windows/Mac may need a third-party program here.  
3. github won't let me attach the result, so I've taken a screenshot. This one is the ANSI 104 layout. Possibly the scaling needs to be corrected to get the right size, it looks like it's coming out too big. I had to set the page size to A2 to fit it in.

This looks like a step in the right direction.
![print2pdf](https://cloud.githubusercontent.com/assets/12673886/9826463/d2163fa4-58da-11e5-9e0a-59b3771c6159.jpg)

**Comment by iandoug** [comment]
Okay I've now got a way to go "real life size" on screen, and if you print to PDF at 100%, you get real life size too. I can send you a PDF of a 60% layout (just because that fits nicely on an A3 printer, which is far more common than a larger format).

**Comment by perpetersen** [comment]
Hi Ian D
I am now relieved from the workload after the IBC show, and I start to be able to see my desk again… Once again, thank you for your efforts with this. If possible, I would really like to see the output print file you are mentioning, but please send it in 100% and not 60%. You can send it to me at per.petersen@logickeyboard.com
